runtime/engine.cpp

```diff
@@ -353,20 +353,39 @@ void api::Engine::abort(const char *reason) { ::abort(CONTEXT, reason); }
 bool api::Engine::eval_toplevel(const char *path, MutableHandleValue result) {
   JSContext *cx = CONTEXT;
   RootedValue ns(cx);
-  if (!scriptLoader->load_top_level_script(path, &ns)) {
+  RootedValue tla_promise(cx);
+  if (!scriptLoader->load_top_level_script(path, &ns, &tla_promise)) {
     return false;
   }
 
+  // Drive TLA as far as possible. Rejections during pre-initialization are
+  // treated as top-level exceptions. TLA may remain unresolved, in which case
+  // it will continue tasks at runtime. Rejections after pre-intialization remain
+  // unhandled rejections for now.
```

Comment:
It'd be good to integrate this with the job queue draining below. Or rather, replace the latter with this wholesale, by unconditionally running `run_event_loop` and only then checking of `tla_promise` is an object, and then removing the `while (js::HasJobsPending(cx))` loop below.